File: Python/147__Insertion_Sort_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(None)
        temp = dummy

        while head:
            while temp.next and head.val > temp.next.val:
                temp = temp.next
            left_unsorted = head.next
            head.next = temp.next
            temp.next = head
            head = left_unsorted
            temp = dummy

        return dummy.next
        

File: Java/147__Insertion_Sort_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode cur = head.next, prev = head, next;

        while (cur != null) {
            next = cur.next;
            if (cur.val >= prev.val) {
                prev = cur;
            } else if (cur.val < head.val) {
                cur.next = head;
                prev.next = next;
                head = cur;
            } else {
                ListNode ptr = head;
                while (cur.val > ptr.next.val) {
                    ptr = ptr.next;
                }
                cur.next = ptr.next;
                ptr.next = cur;
                prev.next = next;
            }
            cur = next;
        }
        return head;
    }
}
